File: integration_tests/kotlin_client_tests_real/client-app/src/test/kotlin/dev/graphlink/test/QueriesTest.kt
package dev.graphlink.test

import dev.graphlink.test.generated.client.GraphLinkClient
import dev.graphlink.test.generated.enums.UserStatus
import dev.graphlink.test.generated.interfaces.SearchResult
import dev.graphlink.test.generated.types.PostResult
import dev.graphlink.test.generated.types.User_IdNameStatus
import dev.graphlink.test.generated.types.UserResult
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Nested
import kotlin.test.*
import dev.graphlink.test.Fixtures.newRealClient

class QueriesTest {

    private lateinit var client: GraphLinkClient

    @BeforeTest
    fun setUp() {
        client = newRealClient()
    }

    // ── fetchUserAndPost — multi-resolver, single round-trip ──────────────────

    @Nested
    inner class FetchUserAndPost {

        @Test
        fun response_isNotNull() = runTest {
            assertNotNull(client.queries.fetchUserAndPost("user-1", "post-1"))
        }

        @Test
        fun user_isNotNull() = runTest {
            assertNotNull(client.queries.fetchUserAndPost("user-1", "post-1").user)
        }

        @Test
        fun user_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.fetchUserAndPost("user-1", "post-1").user.id)
        }

        @Test
        fun user_hasCorrectName() = runTest {
            assertEquals("Alice Smith", client.queries.fetchUserAndPost("user-1", "post-1").user.name)
        }

        @Test
        fun post_isNotNull() = runTest {
            assertNotNull(client.queries.fetchUserAndPost("user-1", "post-1").post)
        }

        @Test
        fun post_hasCorrectId() = runTest {
            assertEquals("post-1", client.queries.fetchUserAndPost("user-1", "post-1").post.id)
        }

        @Test
        fun post_hasCorrectTitle() = runTest {
            assertEquals("Hello World", client.queries.fetchUserAndPost("user-1", "post-1").post.title)
        }

        @Test
        fun post_author_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.fetchUserAndPost("user-1", "post-1").post.author.id)
        }
    }

    // ── fetchUserSummary — partial projection ─────────────────────────────────

    @Nested
    inner class FetchUserSummary {

        @Test
        fun response_isNotNull() = runTest {
            assertNotNull(client.queries.fetchUserSummary("user-1"))
        }

        @Test
        fun getUser_isNotNull() = runTest {
            assertNotNull(client.queries.fetchUserSummary("user-1").getUser)
        }

        @Test
        fun getUser_isProjectedType() = runTest {
            assertIs<User_IdNameStatus>(client.queries.fetchUserSummary("user-1").getUser)
        }

        @Test
        fun getUser_hasCorrectId() = runTest {
            assertEquals("user-1", client.queries.fetchUserSummary("user-1").getUser.id)
        }

        @Test
        fun getUser_hasCorrectName() = runTest {
            assertEquals("Alice Smith", client.queries.fetchUserSummary("user-1").getUser.name)
        }

        @Test
        fun getUser_status_deserializesToEnum() = runTest {
            assertEquals(UserStatus.ACTIVE, client.queries.fetchUserSummary("user-1").getUser.status)
        }
    }

    // ── runSearch — interface dispatch via __typename ─────────────────────────

    @Nested
    inner class RunSearch {

        @Test
        fun results_containUserResult() = runTest {
            val results: List<SearchResult> = client.queries.runSearch("alice").search
            assertNotNull(results)
            assertTrue(results.any { it is UserResult })
        }

        @Test
        fun results_containPostResult() = runTest {
            val results = client.queries.runSearch("alice").search
            assertNotNull(results)
            assertTrue(results.any { it is PostResult })
        }

        @Test
        fun userResult_hasCorrectId() = runTest {
            val results = client.queries.runSearch("alice").search
            val user = results.filterIsInstance<UserResult>().first()
            assertEquals("user-1", user.id)
            assertEquals("Alice Smith", user.name)
        }

        @Test
        fun postResult_hasCorrectId() = runTest {
            val results = client.queries.runSearch("alice").search
            val post = results.filterIsInstance<PostResult>().first()
            assertEquals("post-1", post.id)
        }
    }
}

File: integration_tests/swift_client_tests_real/Tests/GraphLinkGeneratedTests/QueriesTest.swift
import XCTest
@testable import GraphLinkGenerated

// ── fetchUserAndPost — multi-resolver, single round-trip ────────────────────

final class QueriesTest_FetchUserAndPost: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_response_isNotNull() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertNotNil(res)
    }

    func test_user_isNotNull() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertNotNil(res.user)
    }

    func test_user_hasCorrectId() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertEqual(res.user.id, "user-1")
    }

    func test_user_hasCorrectName() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertEqual(res.user.name, "Alice Smith")
    }

    func test_post_isNotNull() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertNotNil(res.post)
    }

    func test_post_hasCorrectId() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertEqual(res.post.id, "post-1")
    }

    func test_post_hasCorrectTitle() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertEqual(res.post.title, "Hello World")
    }

    func test_post_author_hasCorrectId() async throws {
        let res = try await client.queries.fetchUserAndPost(userId: "user-1", postId: "post-1")
        XCTAssertEqual(res.post.author?.id, "user-1")
    }
}

// ── fetchUserSummary — partial projection ────────────────────────────────────

final class QueriesTest_FetchUserSummary: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_response_isNotNull() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertNotNil(res)
    }

    func test_getUser_isNotNull() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertNotNil(res.getUser)
    }

    func test_getUser_isProjectedType() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertTrue(res.getUser is User_IdNameStatus)
    }

    func test_getUser_hasCorrectId() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertEqual(res.getUser.id, "user-1")
    }

    func test_getUser_hasCorrectName() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertEqual(res.getUser.name, "Alice Smith")
    }

    func test_getUser_status_deserializesToEnum() async throws {
        let res = try await client.queries.fetchUserSummary(id: "user-1")
        XCTAssertEqual(res.getUser.status, UserStatus.active)
    }
}

// ── runSearch — interface dispatch via __typename ────────────────────────────

final class QueriesTest_RunSearch: XCTestCase {
    var client: GraphLinkClient!

    override func setUp() {
        client = Fixtures.newRealClient()
    }

    func test_results_containUserResult() async throws {
        let results = try await client.queries.runSearch(term: "alice").search
        XCTAssertNotNil(results)
        XCTAssertTrue(results.contains { $0 is UserResult })
    }

    func test_results_containPostResult() async throws {
        let results = try await client.queries.runSearch(term: "alice").search
        XCTAssertNotNil(results)
        XCTAssertTrue(results.contains { $0 is PostResult })
    }

    func test_userResult_hasCorrectId() async throws {
        let results = try await client.queries.runSearch(term: "alice").search
        let user = results.compactMap { $0 as? UserResult }.first
        XCTAssertNotNil(user)
        XCTAssertEqual(user?.id, "user-1")
        XCTAssertEqual(user?.name, "Alice Smith")
    }

    func test_postResult_hasCorrectId() async throws {
        let results = try await client.queries.runSearch(term: "alice").search
        let post = results.compactMap { $0 as? PostResult }.first
        XCTAssertNotNil(post)
        XCTAssertEqual(post?.id, "post-1")
    }
}
